connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -188,22 +167,25 @@ public void addConnector(ConnectorConfig connConfig, ConnectorContext ctx) {
             throw new ConnectException("Connector with name " + connName + " already exists");
 
         final Connector connector = instantiateConnector(connClass);
+        WorkerConnector workerConnector = new WorkerConnector(connName, connector, ctx, lifecycleListener);
+
         log.info("Instantiated connector {} with version {} of type {}", connName, connector.version(), connClass.getName());
-        connector.initialize(ctx);
+        workerConnector.initialize();
         try {
-            connector.start(connConfig.originalsStrings());
+            workerConnector.start(connConfig.originalsStrings());
         } catch (ConnectException e) {
             throw new ConnectException("Connector threw an exception while starting", e);
         }
 
-        connectors.put(connName, connector);
+        connectors.put(connName, workerConnector);
 
         log.info("Finished creating connector {}", connName);
     }
 
     /* Now that the configuration doesn't contain the actual class name, we need to be able to tell the herder whether a connector is a Sink */
     public boolean isSinkConnector(String connName) {
-        return SinkConnector.class.isAssignableFrom(connectors.get(connName).getClass());
+        WorkerConnector workerConnector = connectors.get(connName);
+        return SinkConnector.class.isAssignableFrom(workerConnector.delegate.getClass());
```

Comment:
Took longer to find than I'd like to admit. I was thinking if there was additional validation we could do, but this check likely would have been the basis for such validation, so I'm not sure it would have helped. Might be worthwhile committing a followup anyway to at least ensure that this code is covered in unit testing.